python-threatexchange/threatexchange/stopncii/api.py

```diff
@@ -0,0 +1,251 @@
+            "Ocp-Apim-Subscription-Key": subscription_key,
+        }
+
+    def _get_session(self):
+        """
+        Custom requests sesson
+
+        Ideally, should be used within a context manager:
+        ```
+        with self._get_session() as session:
+            session.get()...
+        ```
+
+        If using without a context manager, ensure you end up calling close() on
+        the returned value.
+        """
+        session = requests.Session()
+        session.headers.update(self._auth_headers)
+        session.mount(
+            self.BASE_URL,
+            adapter=TimeoutHTTPAdapter(
+                timeout=60,
+                max_retries=Retry(
+                    total=4,
+                    status_forcelist=[429, 500, 502, 503, 504],
+                    allowed_methods=["HEAD", "GET", "OPTIONS"],
+                    backoff_factor=0.2,  # ~1.5 seconds of retries
+                ),
+            ),
+        )
+        return session
+
+    def _get(self, endpoint: str, **json) -> t.Any:
+        """
+        Perform an HTTP GET request, and return the JSON response payload.
+        Same timeouts and retry strategy as `_get_session` above.
+        """
+
+        url = "/".join((self.BASE_URL, self.API_VERSION, endpoint))
+        with self._get_session() as session:
+            response = session.get(url, json=json)
+            response.raise_for_status()
+            return response.json()
+
+    def _post(self, endpoint: str, *, json=None, params=None) -> t.Any:
+        """
+        Perform an HTTP GET request, and return the JSON response payload.
+        Same timeouts and retry strategy as `_get_session` above.
+        """
+
+        url = "/".join((self.BASE_URL, self.API_VERSION, endpoint))
+        with self._get_session() as session:
+            response = session.post(url, json=json, params=params or {})
+            response.raise_for_status()
+            return response.json()
+
+    def fetch_hashes(
+        self, *, start_timestamp: int = None, next_page: str = None
+    ) -> FetchHashesResponse:
+        """
+        Fetch a series of update records from the hash API.
+
+        Records represent the current snapshot of all data, so if you see
+        the same SignalType+Hash in a later iteration, it should completely
+        replace the previously observed record.
+        """
+        if start_timestamp is None:
+            start_timestamp = self.DEFAULT_START_TIMESTAMP
+        params: t.Dict[str, t.Any] = {"startTimestamp": start_timestamp}
+        if next_page is not None:
+            params["nextPageToken"] = next_page
+        json_val = self._get("FetchHashes", **params)
+        logging.debug("StopNCII FetchHashes returns: %s", json_val)
+        return dacite.from_dict(
+            data_class=FetchHashesResponse,
+            data=json_val,
+            config=dacite.Config(cast=[enum.Enum, set]),
+        )
+
+    def fetch_hashes_iter(
```

Comment:
Question: With hashexchange APIs, there must be a translator layer between fb_threatexchange → fetcher and stopncii → fetcher. The layer will have to translate for example -- StopNCIIHashRecord to BankMemberSignal arguments.

Where do you imagine that will happen?

Reply:
Not able to parse your question - the answer is in the existing interfaces. You must implement the interfaces for fetching to work. You can put the translation layer anywhere you want as long as long as the interface returns the right value. Here's the exit of fetch/sync so it must be before/during here https://github.com/facebook/ThreatExchange/blob/main/python-threatexchange/threatexchange/fetcher/fetch_state.py#L267-L269